src/algebra/smul_with_zero.lean

```diff
@@ -157,6 +148,53 @@ def mul_action_with_zero.comp_hom (f : R' →*₀ R) : mul_action_with_zero R' M
   one_smul := λ m, by simp,
   .. smul_with_zero.comp_hom M f.to_zero_hom}
 
+end has_zero
+
+section add_monoid
+variables [add_monoid M] (R M)
+
+/-- Typeclass for a distributive multiplicative action of a monoid with zero `R` on an additive
+monoid `M` such that `0 : R` is left-absorbent and `0 : M` is right-absorbent. -/
+@[ext] class distrib_mul_action_with_zero extends distrib_mul_action R M :=
+-- this field is copied from `mul_action_with_zero`, as `extends` behaves poorly
+(zero_smul : ∀ m : M, (0 : R) • m = 0)
+
+@[priority 100] -- see Note [lower instance priority]
+instance distrib_mul_action_with_zero.to_mul_action_with_zero
+  [m : distrib_mul_action_with_zero R M] :
+  mul_action_with_zero R M :=
+{..m}
+
+variables {R M} [distrib_mul_action_with_zero R M] [add_monoid M'] [has_scalar R M']
+
+/-- Pullback a `distrib_mul_action_with_zero` structure along an injective monoid homomorphism. -/
+@[reducible] -- See note [reducible non-instances]
+protected def function.injective.distrib_mul_action_with_zero (f : M' →+ M) (hf : injective f)
+  (smul : ∀ (a : R) b, f (a • b) = a • f b) :
+  distrib_mul_action_with_zero R M' :=
+{ ..hf.distrib_mul_action f smul, ..hf.mul_action_with_zero (f : zero_hom M' M) smul }
+
+/-- Pushforward a `distrib_mul_action_with_zero` structure along a surjective monoid homomorphism.
+-/
+@[reducible] -- See note [reducible non-instances]
+protected def function.surjective.distrib_mul_action_with_zero (f : M →+ M') (hf : surjective f)
+  (smul : ∀ (a : R) b, f (a • b) = a • f b) :
+  distrib_mul_action_with_zero R M' :=
+{ ..hf.distrib_mul_action f smul, ..hf.mul_action_with_zero (f : zero_hom M M') smul }
+
+variables (M)
+
+/-- Compose a `distrib_mul_action_with_zero` with a `monoid_with_zero_hom`, with action
+`f r' • m`. -/
+def distrib_mul_action_with_zero.comp_hom (f : R' →*₀ R) : distrib_mul_action_with_zero R' M :=
+{ ..distrib_mul_action.comp_hom M f.to_monoid_hom, ..mul_action_with_zero.comp_hom M f }
+
+/-- Each element of the monoid defines an additive monoid homomorphism. -/
```

Comment:
```suggestion
/-- Each element of the monoid defines an additive monoid homomorphism.

This is a stronger version of `distrib_mul_action.to_add_monoid_End`. -/
```

Can you add a comment to the docstring of `module.to_add_monoid_End` referencing this definition too?

I guess it would also be good to have `distrib_mul_action_with_zero.to_module_End`, although I don't know if there's an easy place to put that in the hierarchy. We already have `distrib_mul_action.to_module_End`.